Fixes passing kwargs with unknown keys to Parquet metadata writing fu…

jsnowacki opened the pull request:
Relates to issue #2985. `kwargs` were passed to a functions without `kwargs` support, which caused errors on unknown arguments present in `kwargs`, like `compression` but possibly others. PR adds also a specific test for the issue.

jcrist commented on dask/dataframe/io/parquet.py:
```diff
@@ -413,8 +420,16 @@ def _write_partition_pyarrow(df, open_with, filename, write_index,
+            safe_kwargs = {k: v for k, v in kwargs.items() if k in spec.args}
+            safe_kwargs['schema'] = t.schema
+            safe_kwargs['where'] = fil
+            parquet.write_metadata(**safe_kwargs)
```
These args/kwargs to `write_metadata` change so infrequently (and arrow devs are active enough on dask) that I'd rather not rely on `getargspec`, but instead manually cull them down explicitly in the code.

jsnowacki replied:
As commented in with `test_writing_parquet_with_unknown_kwargs`, in case we'd like to have a fixed number of arguments, it is better to make then into named arguments of function `_write-<engine>`. As for adding just explicit argument names list, it's fine but IMO it creates additional point of change if the underlying engine implementation changes.

jcrist replied:
We use `kwargs` in `to_parquet`/`read_parquet` to specify arguments to the backend engines that differ between them. How the engines use these (via kwargs or explicit keywords) doesn't really matter, the user interface is the intention. In this case you could either:
- Make the `_write_*` take explicit keywords, and explicitly pass them to the underlying functions. This would be nicer, as we can error at call time instead of compute time for invalid parameters.
- Just filter the keyword arguments to `write_metadata`. This could be as simple as:

```python
metadata_kwargs = {'version', 'use_deprecated_int96_timestamps', 'coerce_timestamps'}
parquet.write_metadata(t.schema, fil, **{k: v for k, v in kwargs.items() if k in metadata_kwargs})
```

jsnowacki replied:
Sure, I'll fix it. I'll give the first one a go as it'll be much more verbose for later use.

jcrist commented on dask/dataframe/io/parquet.py:
```diff
@@ -223,11 +223,18 @@ def _write_fastparquet(df, path, write_index=None, append=False,
+        safe_kwargs['object_encoding'] = object_encoding
+        safe_kwargs['index_cols'] = index_cols
+        safe_kwargs['ignore_columns'] = partition_on
+        fmd = fastparquet.writer.make_metadata(**safe_kwargs)
```
Was a change for fastparquet necessary? From looking at the code, it looks like kwargs are only passed to `make_metadata`, which should then error if invalid kwargs are passed in. This is different from the pyarrow case you were fixing, where valid kwargs for the writer were erroneously being passed to the metadata writer.

jsnowacki replied:
The added test `test_writing_parquet_with_unknown_kwargs` makes this function failed as the `unknown_key` is not removed. Indeed, this key is bogus but this will happen for any key we did not anticipate before and it is used in some other function. The code calling this function only pops some variables it uses internally but leaves all the others to this function. In general, see the reasoning in the comment for the test.

jcrist commented on dask/dataframe/io/tests/test_parquet.py:
```diff
@@ -613,3 +613,15 @@ def test_writing_parquet_with_compression(tmpdir, compression, engine):
+                       'y': [1, 2, 3] * 10})
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    ddf.to_parquet(fn,  engine=engine, unknown_key='unknown_value')
```
This should error imo. If the user passes in unknown invalid keyword arguments, they shouldn't be silently ignored. What shouldn't fail is passing in valid keyword arguments like `compression` or `use_dictionary` (for pyarrow).

jsnowacki replied:
This would require `kwargs` content validation using all the arguments of all the functions used inside the write function, which is impractical and I've never seen it done, as the point of `kwargs` is to have arguments that are not explicitly known, because there are used by some other function. The problem here is that we mix functions like `write_table`, which has `kwargs` and the unknown ones are ignored, and `write_metadata`, which doesn't have `kwargs` and, thus, fails if it gets something unknown in `kwargs`. For the latter case we should only extract required arguments from `kwargs` and ignore the rest as they may be used by some other function, which we don't know, if we decide to use `kwargs`. If we want to restrict the number of arguments used for the particular engine implementation, we should remove `kwargs` altogether from the `_write_<engine>` function and keep only the known ones. This will fail as intended on `write` call in `to_parquet` function; see: https://github.com/jsnowacki/dask/blob/f054102e35b550670052ed3a00b8dcbae0d3ecb1/dask/dataframe/io/parquet.py#L626. Note that using just specific arguments in the write function for the engine will create a new place that any change to the engine has to be implemented as well.

jsnowacki commented:
Following the comments, I've removed `kwargs` from `_write_<engine>` functions and replaced them with engine-specific named arguments. While there are still `kwargs` in `write_parquet` function, if some inappropriate arguments are passed, the function fails; I've added test for that. Also, I included `NotImplementedError` for `ignore_divisions` as it was dangling without any information.

jsnowacki commented:
BTW I've noticed that on Windows the following test fail mostly because of different handling of them on Windows:

- `dask.dataframe.io.tests.test_parquet.test_read_glob.(fastparquet-fastparquet)` (`dir` columns are added to the read DF, which is not expected in the test)
- `dask.dataframe.io.tests.test_parquet.test_read_glob.(pyarrow-fastparquet)`  (it raises `AttributeError("'ParquetFile' object has no attribute '_schema'")` but it seems to be related wit not finding `_metadata`)
- `dask.dataframe.io.tests.test_parquet.test_append_with_partition` (different types `int32` and `int64`, but should be the same)
- `dask.dataframe.io.tests.test_parquet.test_drill_scheme` (because the values are in the last column called `dir`, not `dir0`)

martindurant commented:
@jsnowacki , I'd be interested to know whwther https://github.com/dask/fastparquet/pull/232 solves any of those issues.

jcrist commented on dask/dataframe/io/parquet.py:
```diff
@@ -375,13 +373,22 @@ def _read_pyarrow_parquet_piece(open_file_func, piece, columns, index_cols,
+                   use_dictionary=True, compression='snappy',
+                   use_deprecated_int96_timestamps=None,
+                   coerce_timestamps=None,
+                   flavor=None, chunk_size=None):
```
The copying of the default values here is rather unfortunate, and fragile to changes to keyword default values. On second thought, I think I prefer my second solution I suggested previously. This diff seems to do the correct thing:

```diff
diff --git a/dask/dataframe/io/parquet.py b/dask/dataframe/io/parquet.py
index 36dc347..d4acf17 100644
--- a/dask/dataframe/io/parquet.py
+++ b/dask/dataframe/io/parquet.py
@@ -373,6 +373,14 @@ def _read_pyarrow_parquet_piece(open_file_func, piece, columns, index_cols,
         return df
 
 
+_write_table_kwargs = {'row_group_size', 'version', 'use_dictionary',
+                       'compression', 'use_deprecated_int96_timestamps',
+                       'coerce_timestamps', 'flavor', 'chunk_size'}
+
+_write_metadata_kwargs = {'version', 'use_deprecated_int96_timestamps',
+                          'coerce_timestamps'}
+
+
 def _write_pyarrow(df, path, write_index=None, append=False,
                    ignore_divisions=False, partition_on=None,
                    storage_options=None, **kwargs):
@@ -384,6 +392,11 @@ def _write_pyarrow(df, path, write_index=None, append=False,
         raise NotImplementedError("`partition_on` not implemented for "
                                   "`engine='pyarrow'`")
 
+    if set(kwargs).difference(_write_table_kwargs):
+        msg = ("Unexpected keyword arguments: "
+               "%r" % list(set(kwargs).difference(_write_table_kwargs)))
+        raise TypeError(msg)
+
     if write_index is None and df.known_divisions:
         write_index = True
 
@@ -413,8 +426,9 @@ def _write_partition_pyarrow(df, open_with, filename, write_index,
 
     if metadata_path is not None:
         with open_with(metadata_path, 'wb') as fil:
-            kwargs.pop('compression', None)
-            parquet.write_metadata(t.schema, fil, **kwargs)
+            kwargs2 = {k: v for k, v in kwargs.items()
+                       if k in _write_metadata_kwargs}
+            parquet.write_metadata(t.schema, fil, **kwargs2)
 
 
 # ----------------------------------------------------------------------
```

jcrist replied:
A second benefit of doing the above is that we only encode user specified keywords in the tasks, rather than every value, making the tasks mildly smaller.

jcrist commented on dask/dataframe/io/tests/test_parquet.py:
```diff
@@ -613,3 +613,14 @@ def test_writing_parquet_with_compression(tmpdir, compression, engine):
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    with pytest.raises(TypeError):
+        ddf.to_parquet(fn,  engine=engine, unknown_key='unknown_value')
```
This only tests passing in an unexpected keyword argument, which is a different issue than the one you experienced, with passing in backend-specific keyword arguments. Could you add a test with something specific to each backend (e.g. `use_dictionary` for pyarrow)?

jsnowacki commented:
@martindurant I've tried your branch and it fixes: `dask.dataframe.io.tests.test_parquet.test_read_glob.(fastparquet-fastparquet)` and `dask.dataframe.io.tests.test_parquet.test_drill_scheme`.

jsnowacki commented:
I've incorporated the suggested changes. Also, I've added a test for known and unknown `kwargs`.

jcrist commented:
Failing test is unrelated. This looks good to me - thanks @jsnowacki! Merging.
